src/libraries/System.DirectoryServices.Protocols/src/System/DirectoryServices/Protocols/common/DirectoryControl.cs

```diff
@@ -132,102 +154,165 @@ public virtual byte[] GetValue()
 
         internal static void TransformControls(DirectoryControl[] controls)
         {
-            for (int i = 0; i < controls.Length; i++)
+            Span<byte> attributeNameScratchSpace = stackalloc byte[AttributeNameStackAllocationThreshold];
+
+            try
             {
-                Debug.Assert(controls[i] != null);
-                byte[] value = controls[i].GetValue();
-                if (controls[i].Type == "1.2.840.113556.1.4.319")
+                for (int i = 0; i < controls.Length; i++)
                 {
-                    // The control is a PageControl.
-                    object[] result = BerConverter.Decode("{iO}", value);
```

Comment:
While looking to see if there are any interesting edge cases popping up from this, I found a case where we're increasing the range of supported values, in a surprising way.

The DER sequence `30 05 02 01 03 04 00` is a `SEQUENCE(INTEGER(3), OCTET_STRING(''))`.  Under BER (and CER) it can (or, for CER, must) also be encoded with an indefinite length: `30 80 02 01 03 04 00 00 00` (change the length of the sequence from 0x05 to 0x80 and add two 0x00 values on the end)

I was incredibly surprised to learn that `BerConverter.Decode` doesn't seem to support indefinite-length encoding.

If it's important to continue NOT supporting indefinite-length encoding, that will require special handling.

The main thing I was trying to test for is "does the existing support trailing data", since the new code seems to.  And it looks like yes (as measured from the perspective of BerDecoder).

Reply:
That's caught me by surprise too, thanks for pointing it out. `BerConverter.Decode`'s lack of support for indefinite-length encoding is consistent between Windows and Linux - they may have interpreted [section 5.1 of RFC4511](https://datatracker.ietf.org/doc/html/rfc4511#section-5.1), which states that LDAP shall be encoded for exchange using BER and only use the definite form of length encoding, in order to ease the encoding/decoding overhead. Since their `ber_scanf` interpretation knows it'll never need to deal with indefinite length encoding, it's just not supported.

I don't think it precludes us from using an ASN.1 parser which supports it. Anyone who sends this is breaching the protocol specification. Section 4.1.1 notes:
> If the server receives an LDAPMessage from the client in which the
> LDAPMessage SEQUENCE tag cannot be recognized, the messageID cannot
> be parsed, the tag of the protocolOp is not recognized as a request,
> or the encoding structures or lengths of data fields are found to be
> incorrect, then the server SHOULD return the Notice of Disconnection
> described in [Section 4.4.1](https://datatracker.ietf.org/doc/html/rfc4511#section-4.4.1), with the resultCode set to protocolError,
> and MUST immediately terminate the LDAP session as described in
> [Section 5.3](https://datatracker.ietf.org/doc/html/rfc4511#section-5.3).
>
> In other cases where the client or server cannot parse an LDAP PDU,
> it SHOULD abruptly terminate the LDAP session ([Section 5.3](https://datatracker.ietf.org/doc/html/rfc4511#section-5.3)) where
> further communication (including providing notice) would be
> pernicious.  Otherwise, server implementations MUST return an
> appropriate response to the request, with the resultCode set to
> protocolError.

As I read it: if the client receives an LDAP PDU containing an indefinite-length BER-encoded value from the server but can still parse this PDU without any errors, the LDAP specification doesn't explicitly define the expected client-side action.

I've read the intent of the RFC's restriction as being solely to reduce encoding/decoding overhead, so unless you'd prefer to flag the server's breach of RFC4511 with an exception, I'm comfortable with `DirectoryControl.TransformControls` silently parsing any indefinite-length BER value properly and proceeding as it currently does.